NDTensors/src/dense/dense.jl

```diff
@@ -20,38 +20,43 @@ struct Dense{ElT,VecT<:AbstractArray} <: TensorStorage{ElT}
   end
 end
 
-function Dense(data::VecT) where {VecT<:AbstractArray{ElT}} where {ElT}
-  return Dense{ElT,VecT}(data)
+## We need a general typeinfo constructor for Dense
+default_Densetype(datatype::Type{<:AbstractArray{ElT}}) where {ElT} = default_storagetype(datatype)
+default_Densetype(eltype::Type{<:Number}) = default_Densetype(default_datatype(eltype))
+default_Densetype() = default_Densetype(default_datatype(default_eltype()))
+## End typeinfo based constructors 
+
+#Start with high information constructors and move to low information constructors
+function Dense{ElT, VecT}() where {ElT, VecT<:AbstractArray{ElT}}
+  return default_Densetype(VecT)(VecT())
 end
 
-function Dense(data::Array{ElT}) where {ElT}
-  return Dense{ElT,Vector{ElT}}(vec(data))
+# Construct from a set of indices
+function Dense{ElT,VecT}(inds::Tuple) where {ElT,VecT<:AbstractArray{ElT}}
+  return default_Densetype(VecT)(VecT(undef, dim(inds)))
+  #eturn Dense(VecT(undef, dim(inds)))
 end
 
+# This function is ill-defined. It cannot transform a complex type to real...
 function Dense{ElR}(data::AbstractArray{ElT}) where {ElR,ElT}
-  return ElT == ElR ? Dense(data) : Dense(ElR.(data))
+  return default_Densetype(ElR)(data)
```

Comment:
Again, I don't think `default_Densetype` is necessary here, this is better handled by a type conversion or `adapt`.

Reply:
We should think about the `default_storagetype` function as only used when we don't have enough information based on the inputs about the full storage type, which is not the case here (since this construct a storage type `Dense{ElR,AbstractArray{ElR}}` which is fully inferrable from the inputs).

Reply:
Ahh, I see!